### PR TITLE
Feature/rpc temporal service

### PR DESCRIPTION
**Pull request by 48d90782**
Add temporal RPC implementation
Add wrappers for the temporal API (not all, but for the rest, we should figure out the responses payload)

### CODE REVIEW

**Comment by 48d90782** [comment]
resolves #8

File: plugins/temporal/temporal.go [hunk 1]
```diff
@@ -46,9 +59,11 @@ func (p *Provider) Serve() chan error {
 	p.serviceClient, err = client.NewClient(client.Options{
 		HostPort:      p.config.Address,
 		Namespace:     p.config.Namespace,
-		DataConverter: NewRRDataConverter(),
+		DataConverter: roadrunner_temporal.NewRRDataConverter(),
```
Comment (wolfy-j) [inline plugins/temporal/temporal.go:62]: We need to improve it to pack into Payloads directly.

File: plugins/temporal/temporal.go [hunk 2]
```diff
@@ -46,9 +59,11 @@ func (p *Provider) Serve() chan error {
...
 	})
 
+	//p.Rpc.client = p.serviceClient
```
Comment (wolfy-j) [inline plugins/temporal/temporal.go:65]: remove

File: plugins/temporal/temporal.go [hunk 3]
```diff
@@ -17,13 +19,20 @@ type Config struct {
...
 type Temporal interface {
 	GetClient() (client.Client, error)
+	GetConfig() Config
 	CreateWorker(taskQueue string, options worker.Options) (worker.Worker, error)
 }
 
 type Provider struct {
+	endure.Named
+	// Rpc implementation
+	Rpc
```
Comment (wolfy-j) [inline plugins/temporal/temporal.go:29]: where is the interface?

File: plugins/temporal/rpc.go [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+- the method is exported.
+- the method has two arguments, both exported (or builtin) types.
+- the method's second argument is a pointer.
+- the method has return type error.
+*/
+type Rpc struct {
+	client client.Client
+}
+
+func (p *Provider) Name() string {
```
Comment (wolfy-j) [inline plugins/temporal/rpc.go:73]: does it make sense to move it into temporal.go file?

File: plugins/temporal/rpc.go [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+// NOTE: if the started workflow return ContinueAsNewError during the workflow execution, the
+// return result of GetRunID() will be the started workflow run ID, not the new run ID caused by ContinueAsNewError,
+// however, Get(ctx context.Context, valuePtr interface{}) will return result from the run which did not return ContinueAsNewError.
+// Say ExecuteWorkflow started a workflow, in its first run, has run ID "run ID 1", and returned ContinueAsNewError,
+// the second run has run ID "run ID 2" and return some result other than ContinueAsNewError:
+// GetRunID() will always return "run ID 1" and  Get(ctx context.Context, valuePtr interface{}) will return the result of second run.
+// NOTE: DO NOT USE THIS API INSIDE A WORKFLOW, USE workflow.ExecuteChildWorkflow instead
+
+type ExecuteWorkflowIn struct {
+	Options client.StartWorkflowOptions `json:"options"`
```
Comment (wolfy-j) [inline plugins/temporal/rpc.go:115]: i wonder if we really need to define name mapping since we still use native structures

Comment (wolfy-j) [reply plugins/temporal/rpc.go:115]: i mean it's not going to be consistent

File: plugins/temporal/rpc.go [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+// - workflowID is required.
+// - runID can be default(empty string). if empty string then it will pick the running execution of that workflow ID.
+// - queryType is the type of the query.
+// - args... are the optional query parameters.
+// The errors it can return:
+//  - BadRequestError
+//  - InternalServiceError
+//  - EntityNotExistError
+//  - QueryFailError
+func (r *Rpc) QueryWorkflow(in QueryWorkflowIn, out *payload.RRPayload) error {
```
Comment (wolfy-j) [inline plugins/temporal/rpc.go:456]: need to double check that this payload properly represents all responses

File: plugins/temporal/rpc.go [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+	if err != nil {
+		return err
+	}
+
+	_ = resp
+	return nil
+}
+
+// Close client and clean up underlying resources.
+func (r *Rpc) Close(_ EmptyStruct, out *EmptyStruct) error {
```
Comment (wolfy-j) [inline plugins/temporal/rpc.go:489]: remove it, you can't close client over RPC

File: plugins/temporal/rpc.go [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+	}
+	return nil
+}
+
+type DescribeWorkflowExecutionIn struct {
+	WorkflowId    string `json:"wid"`
+	WorkflowRunId string `json:"rid"`
+}
+
+func (r *Rpc) DescribeWorkflowExecution(in DescribeWorkflowExecutionIn, ) error {
```
Comment (wolfy-j) [inline plugins/temporal/rpc.go:476]: drop it for now

**Review by wolfy-j — CHANGES_REQUESTED** [review]

File: plugins/temporal/temporal.go (hunk 1 above)
Comment (48d90782) [reply plugins/temporal/temporal.go:62]: Agree, will do in a separate ticket

File: plugins/temporal/temporal.go (hunk 2 above)
Comment (48d90782) [reply plugins/temporal/temporal.go]: done

File: plugins/temporal/rpc.go (hunk 4 above)
Comment (48d90782) [reply plugins/temporal/rpc.go:73]: sure, it does. done

File: plugins/temporal/temporal.go (hunk 3 above)
Comment (48d90782) [reply plugins/temporal/temporal.go]: yeah, we do not need here a Named interface and RPC struct integration. The Provider struct implements roadrunner.rpc.Plugin interface.

File: plugins/temporal/rpc.go (hunk 6 above)
Comment (48d90782) [reply plugins/temporal/rpc.go:440]: Agree, will be covered with tests

File: plugins/temporal/rpc.go (hunk 8 above)
Comment (48d90782) [reply plugins/temporal/rpc.go]: done

File: plugins/temporal/rpc.go (hunk 7 above)
Comment (48d90782) [reply plugins/temporal/rpc.go:474]: ok, removed

File: plugins/temporal/rpc.go (hunk 5 above)
Comment (48d90782) [reply plugins/temporal/rpc.go:100]: This is an issue. Should be our StartWorkflowOptions with our JSON tags. Fixed.

**Review by wolfy-j — APPROVED** [review]

**Pull request merged**